language/move-prover/boogie-backend/src/spec_translator.rs

```diff
@@ -57,7 +57,7 @@ pub struct SpecTranslator<'env> {
     /// choice function. If an expression is duplicated and then later specialized by a type
     /// instantiation, it will have a different node id, but again the same instantiations
     /// map to the same node id, which is the desired semantics.
-    lifted_choice_infos: Rc<RefCell<HashMap<ExpData, LiftedChoiceInfo>>>,
+    lifted_choice_infos: Rc<RefCell<HashMap<(ExpData, Vec<Type>), LiftedChoiceInfo>>>,
```

Comment:
Did you try to just use NodeId instead of the full ExpData? Usage of ExpData here was always kind of strange.

I'm fine to move forward with this PR anyway to unblock other work, so we can do in a 2nd PR.

Reply:
I have tried using NodeId but it does not work for the example in Meng's PR

Reply:
Ok, then perhaps just fix the need of &mut and use the way how forks with type_inst are already handled in this file.